sources/vidplay.py

```diff
@@ -30,25 +25,72 @@ def get_vidplay_subtitles(url_data: str) -> Dict:
         
         return {}
 
+    @staticmethod
+    def encode(key: str, v_id: str) -> str:
+        decoded_id = Utilities.decode_data(key, v_id)
+        
+        encoded_base64 = base64.b64encode(decoded_id)
+        decoded_result = encoded_base64.decode("utf-8")
+
+        return decoded_result.replace("/", "_").replace("+", "-")
+
+
     @staticmethod
     def encode_id(v_id: str) -> str:
+        key = VidplayExtractor.get_encryption_key()
+        return VidplayExtractor.encode(key, v_id)
+
+    @staticmethod
+    def encode_embed_id(v_id: str) -> str:
+        key = VidplayExtractor.get_embed_encryption_key()
+        return VidplayExtractor.encode(key, v_id)
+
+    @staticmethod
+    def encode_h(v_id: str) -> str:
+        key = VidplayExtractor.get_h_encryption_key()
+        return VidplayExtractor.encode(key, v_id)
+
+    @staticmethod
+    def get_key(enc: bool, num: int) -> str:
         req = requests.get(VidplayExtractor.KEY_URL)
 
         if req.status_code != 200:
             raise CouldntFetchKeys("Failed to fetch decryption keys!")
-        
+
         matches = re.search(r"\"rawLines\":\s*\[\"(.+)\"\]", req.text)
         if not matches:
             raise CouldntFetchKeys("Failed to extract rawLines from keys page!")
 
-        key1, key2 = json.loads(matches.group(1).replace("\\", ""))
-        decoded_id = Utilities.decode_data(key1, v_id)
-        encoded_result = Utilities.decode_data(key2, decoded_id)
-        
-        encoded_base64 = base64.b64encode(encoded_result)
-        decoded_result = encoded_base64.decode("utf-8")
+        keys = json.loads(matches.group(1).replace("\\", ""))
+        return keys["encrypt" if enc else "decrypt"][num]
+
+    @staticmethod
+    def get_encryption_key() -> str:
```

Comment:
These functions are requesting `VidplayExtractor.KEY_URL` every time `get_key` is called, we should instead fetch all keys once in `vidsrc.py` and pass the keys through.

Reply:
Okay, I tried my best. The problem is that in the code and so in the vidsrc-keys repo, there is a bit of confusion. It seems that all keys are for vidsrc. In the future, I would suggest dividing them based on the website (i.e. vidsrc and f2cloud), so that it would be possible to write a cleaner code.